auth/src/main/java/org/aerogear/auth/impl/OIDCAuthCodeImpl.java

```diff
@@ -20,6 +43,120 @@ public OIDCAuthCodeImpl(final AuthServiceConfig config) {
      */
     @Override
     public Principal authenticate(final ICredential credential) {
-        throw new IllegalStateException("Not implemented");
+        UserPrincipalImpl user = null;
+        try {
+            userIdentity = getIdentityInformation(credential);
+            user = UserPrincipalImpl
+                    .newUser()
+                    .withAuthenticator(this)
+                    .withUsername(parseUsername())
+                    .withCredentials(credential)
+                    .withEmail(parseEmail())
+                    .withRoles(parseRoles())
+                    .build();
+        } catch (JSONException e) {
+            e.printStackTrace();
+        }
+    return user;
+    }
+
+    private String parseUsername() throws JSONException {
+        String username = "Unknown Username";
+        if (userIdentity != null) {
+            // get the users username
+            if (userIdentity.has(USERNAME) && userIdentity.getString(USERNAME).length() > 0) {
+                username = userIdentity.getString(USERNAME);
+            }
+        }
+        return username;
+    }
+
+    private String parseEmail() throws JSONException {
+        String emailAddress = "Unknown Email";
+        if (userIdentity != null) {
+            // get the users email
+            if (userIdentity.has(EMAIL) && userIdentity.getString(EMAIL).length() > 0) {
+                emailAddress = userIdentity.getString(EMAIL);
+            }
+        }
+        return emailAddress;
+    }
+
+    private Map<RoleKey, IRole> parseRoles() throws JSONException {
```

Comment:
Is `IRole` an interface? I could be wrong but i thought the `I` prefix for interfaces is more of a .Net thing and not common in Java. @ziccardi @matzew what do you think?

Reply:
Lately, the idea is that the prefix should not be used, you are right. It used to be very common in the past, even in the Java World. 
However, it really depends on whether we have or not a standard naming convention in aerogear.
IMHO I think that the 'I' prefix for Interfaces is very useful.

Other opensource software like eclipse uses it extensively. An example: http://git.eclipse.org/c/actf/org.eclipse.actf.common.git/tree/plugins/org.eclipse.actf.core/src/org/eclipse/actf/core/adapt

@matzew WDYT?